This lets the root module provide the LLVM toolchain while allowing non-root
modules to reference the LLVM extension. Buildbarn's shared reformat target
uses clang-format through toolchains_llvm, so this is needed for
@com_github_buildbarn_bb_storage//tools:reformat to resolve its formatter tools
under Bzlmod.
Copied from bb-remote-execution:
https://github.com/buildbarn/bb-remote-execution/blob/3afb89222c2dd68a6f8f71b83a6b59d48e21046a/patches/toolchains_llvm/non-root.diff

diff --git toolchain/extensions/llvm.bzl toolchain/extensions/llvm.bzl
index 2814871..767ad9d 100644
--- toolchain/extensions/llvm.bzl
+++ toolchain/extensions/llvm.bzl
@@ -50,10 +50,10 @@ def _constraint_dict(tags, name):
 
 def _llvm_impl_(module_ctx):
     for mod in module_ctx.modules:
-        if not mod.is_root:
-            fail("Only the root module can use the 'llvm' extension")
         toolchain_names = []
         for toolchain_attr in mod.tags.toolchain:
+            if not mod.is_root:
+                fail("Only the root module can use the 'llvm.toolchain()' tag")
             name = toolchain_attr.name
             toolchain_names.append(name)
             attrs = {
